ntoskrnl/ke/i386/mproc.c

```diff
@@ -0,0 +1,193 @@
+
+        KiSetGdtDescriptorBase(KiGetGdtEntry(&APInfo->Gdt, KGDT_R0_PCR), (ULONG_PTR)&APInfo->Pcr);
+        KiSetGdtDescriptorBase(KiGetGdtEntry(&APInfo->Gdt, KGDT_DF_TSS), (ULONG_PTR)&APInfo->TssDoubleFault);
+        KiSetGdtDescriptorBase(KiGetGdtEntry(&APInfo->Gdt, KGDT_NMI_TSS), (ULONG_PTR)&APInfo->TssNMI);
+
+        KiSetGdtDescriptorBase(KiGetGdtEntry(&APInfo->Gdt, KGDT_TSS), (ULONG_PTR)&APInfo->Tss);
+        // Clear TSS Busy flag (aka set the type to "TSS (Available)")
+        KiGetGdtEntry(&APInfo->Gdt, KGDT_TSS)->HighWord.Bits.Type = 0b1001;
+
+        APInfo->TssDoubleFault.Esp0 = (ULONG_PTR)&APInfo->NMIStackData;
+        APInfo->TssDoubleFault.Esp = (ULONG_PTR)&APInfo->NMIStackData;
+
+        APInfo->TssNMI.Esp0 = (ULONG_PTR)&APInfo->NMIStackData;
+        APInfo->TssNMI.Esp = (ULONG_PTR)&APInfo->NMIStackData;
+
+        // Push LOADER_BLOCK on stack as a parameter
+        KernelStack = (PVOID)((ULONG_PTR)KernelStack - sizeof(PVOID));
+        *(PVOID *)KernelStack = KeLoaderBlock;
+
+        // Push NULL address on stack as a "return" addr
+        KernelStack = (PVOID)((ULONG_PTR)KernelStack - sizeof(PVOID));
+        *(PVOID *)KernelStack = NULL;
+        
+        // Fill the processor state
+        PKPROCESSOR_STATE ProcessorState = &APInfo->Pcr.Prcb->ProcessorState;
+        RtlZeroMemory(ProcessorState, sizeof(*ProcessorState));
+
+        ProcessorState->SpecialRegisters.Cr0 = __readcr0();
+        ProcessorState->SpecialRegisters.Cr3 = __readcr3();
+        ProcessorState->SpecialRegisters.Cr4 = __readcr4();
+
+        ProcessorState->ContextFrame.SegCs = KGDT_R0_CODE;
+        ProcessorState->ContextFrame.SegDs = KGDT_R3_DATA;
+        ProcessorState->ContextFrame.SegEs = KGDT_R3_DATA;
+        ProcessorState->ContextFrame.SegSs = KGDT_R0_DATA;
+        ProcessorState->ContextFrame.SegFs = KGDT_R0_PCR;
+
+        ProcessorState->SpecialRegisters.Gdtr.Base = (ULONG_PTR)APInfo->Gdt;
+        ProcessorState->SpecialRegisters.Gdtr.Limit = sizeof(APInfo->Gdt) - 1;
+        ProcessorState->SpecialRegisters.Idtr.Base = (ULONG_PTR)APInfo->Idt;
+        ProcessorState->SpecialRegisters.Idtr.Limit = sizeof(APInfo->Idt) - 1;
+
+        ProcessorState->SpecialRegisters.Tr = KGDT_TSS;
+
+        ProcessorState->ContextFrame.Esp = (ULONG_PTR)KernelStack;
+        ProcessorState->ContextFrame.Eip = (ULONG_PTR)KiSystemStartup;
+        ProcessorState->ContextFrame.EFlags = __readeflags() & ~EFLAGS_INTERRUPT_MASK;
+
+        // Prepare the LOADER_PARAMETER_BLOCK structure
+        KeLoaderBlock->KernelStack = (ULONG_PTR)KernelStack;
+        KeLoaderBlock->Prcb = (ULONG_PTR)&APInfo->Pcr.Prcb;
+        KeLoaderBlock->Thread = (ULONG_PTR)&APInfo->Pcr.Prcb->IdleThread;
+
+        // Start the CPU
+        DPRINT("Starting CPU: #%u\n", ProcessorCount);
+        if (!HalStartNextProcessor(KeLoaderBlock, ProcessorState))
+        {
+            break;
+        }
+
+        while (READ_PORT_ULONG(&KeLoaderBlock->Prcb) != 0)
```

Comment:
Can you add a comment here to explain what this I/O port is about?

Reply:
Yeah.

And in another note I don’t know if I agree with having multiple processor access the same value in the same instance to check for success,

this will be written and then read in the same instance by two different processors respectively,

maybe someone has an idea for a better signal we can do to tell the BSP the currently spinning up AP is good to go?

Reply:
I think the real quesiton is this: the Prcb in KeLoaderBlock is meant to be a **pointer** to a processor control region block (like those other Process and Thread values that are pointers to KPROCESS and KTHREAD), so, why would you act with an I/O port operation on Prcb?

Reply:
If you want to check whether the Prcb pointer value changes in a atomic way, it's with an InterlockedCompareExchange or similar.

Reply:
`READ_PORT_ULONG` shouldn't be needed, when you use `KeMemoryBarrier()`. But you need another `KeMemoryBarrier();` before the loop. The problem currently is that `KeLoaderBlock->Prcb` is only set to NULL in `KiInitializeKernel`, but the AP will not get there.

Reply:
Yeah I think the point of READ_PORT_ULONG is basically "read once, don't optimize" (like Linux's READ_ONCE).